src/domain/trader.ts

```diff
@@ -0,0 +1,76 @@
+import { BuyEvent, CreateEvent, EventType, Position, SellEvent, TRADER_EVENTS } from '@server/events.types';
+import { v4 as uuidv4 } from 'uuid';
+
+export class Trader {
+    public id: string;
+    name: string;
+    positionMap: Map<string, Object>;
+
+    constructor(name: string) {
+        this.name = name;
+        this.id = uuidv4();
+
+        const event: CreateEvent = {
+            type: EventType.CREATE,
+            name,
+            id: this.id
+        }
+        TRADER_EVENTS[this.id] = [event];
+    }
+
+    buy(symbol: string, date: Date, quantity: number, price: number) {
+        const event: BuyEvent = {
+            type: EventType.BUY,
+            id: this.id,
+            symbol,
+            quantity,
+            date,
+            price,
+        }
+        TRADER_EVENTS[this.id].push(event);
+    }
+
+    sell(symbol: string, date: Date, quantity: number, price: number) {
+        const event: SellEvent = {
+            type: EventType.SELL,
+            id: this.id,
+            symbol,
+            quantity,
+            date,
+            price,
+        }
+        TRADER_EVENTS[this.id].push(event);
+
+        const { positions } = this.positions();
+        const lastTotalPL = positions[positions.length - 1].total_pl;
+
+        if (lastTotalPL - (quantity * price) < 0) {
+            throw new Error('Selling will make position go into the negative');
+        }
```

Comment:
We would want to do this check before pushing to the event stream / persisting the event. Only persist valid state changes